src/simple_uds.rs

```
@@ -0,0 +1,179 @@
+    pub sockpath: path::PathBuf,
+    /// The read and write timeout to use
+    pub timeout: Option<time::Duration>,
+}
+
+impl UdsTransport {
+    /// Create a new UdsTransport without timeouts to use
+    pub fn new<P: AsRef<path::Path>>(sockpath: P) -> UdsTransport {
+        UdsTransport {
+            sockpath: sockpath.as_ref().to_path_buf(),
+            timeout: None,
+        }
+    }
+
+    fn request<R>(&self, req: impl serde::Serialize) -> Result<R, Error>
+    where
+        R: for<'a> serde::de::Deserialize<'a>,
+    {
+        let mut sock = UnixStream::connect(&self.sockpath)?;
+        sock.set_read_timeout(self.timeout)?;
+        sock.set_write_timeout(self.timeout)?;
+
+        serde_json::to_writer(&mut sock, &req)?;
+
+        // NOTE: we don't check the id there, so it *must* be synchronous
+        let resp: R = serde_json::Deserializer::from_reader(&mut sock)
+            .into_iter()
+            .next()
+            .ok_or(Error::Timeout)??;
+        Ok(resp)
+    }
+}
+
+impl Transport for UdsTransport {
+    fn send_request(&self, req: Request) -> Result<Response, ::Error> {
+        Ok(self.request(req)?)
+    }
+
+    fn send_batch(&self, reqs: &[Request]) -> Result<Vec<Response>, ::Error> {
+        Ok(self.request(reqs)?)
+    }
+
+    fn fmt_target(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.sockpath.to_string_lossy())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    #[cfg(not(windows))]
+    use std::os::unix::net::UnixListener;
+    #[cfg(windows)]
+    use uds_windows::UnixListener;
+
+    use std::{
+        fs,
+        io::{Read, Write},
+        process, thread,
+    };
+
+    use super::*;
+    use Client;
+
+    // Test a dummy request / response over an UDS
+    #[test]
+    fn sanity_check_uds_transport() {
+        let socket_path: path::PathBuf = format!("uds_scratch_{}.socket", process::id()).into();
+        // Any leftover?
+        fs::remove_file(&socket_path).unwrap_or_else(|_| ());
```

Not important, but I usually just `let _ = failable_operation();` when I want to ignore errors.